Load AccucCor data into MsRun, PeakGroup, PeakData

jcmatese opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

add researcher to MsRun model
tweak labeled_element_count validator in PeakData model
add command to load_accucor_msruns
add AccuCorDataLoader to utils
add openpyxl to requirement for Excel reading via pandas

## Affected Issue Numbers

- Resolves #45 

## Code Review Notes

A lot of the is new to me, so everything deserves some some close attention.
Attempt to use `transaction` and  to accept load-all or load-nothing.  Been tested some, but not exhaustive.
Some hard-cording of data files expectations (column numbers), that probably is not future-or-past-proof (tuned to the example file).
I did not implement too many post load tests, but I did try and perform some basic validations prior to load.

**EDIT:** We believe the known issue documented below may have been fixed with @lparsons later [model edit](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/55/commits/b7155eaa5a02c4dfcf5c30db0c3b421619126eac)

**Known issue**: It appears that the method for inserting uniquely into MsRun could be tightened up, probably because of time formatting.
example duplicate load:
```
{'_state': <django.db.models.base.ModelState object at 0x7fe5d820d760>,
 'date': datetime.datetime(2021, 4, 28, 17, 1, 12, 215368, tzinfo=<UTC>),
 'id': 560,
 'protocol_id': 1,
 'researcher': 'mneistat',
 'sample_id': 87}
{'_state': <django.db.models.base.ModelState object at 0x7fe5d820d370>,
 'date': datetime.datetime(2021, 4, 28, 17, 39, 57, 605820, tzinfo=<UTC>),
 'id': 617,
 'protocol_id': 1,
 'researcher': 'mneistat',
 'sample_id': 87}
```
resulted from
`python manage.py load_accucor_msruns --accucor_filename "DataRepo/example_data/obob_maven_6eaas_inf.xlsx" --protocol 1 --date "2021-04-23" --researcher "mneistat"`

I must not be setting/formatting the date correct from args.


## Checklist

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

jcmatese commented:
I suspect the the `date` issue is due to some `auto` behavior in the model 

https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/2f18245b9a9c151ed8fb0bd7f8244fca3ec7b66b/DataRepo/models.py#L154

where I would prefer it to be "date the run was performed", not "date the record was entered", but that raised other issues of provenance and database auditing that we have not discussed in detail.

I give up no the linting issues, because Django migrations and Flake8 will always run afoul of each other, when it comes to line length.

lparsons commented:
@jcmatese Looks like there is a missing migration (forgot to check it in perhaps?)

> I give up on the linting issues, because Django migrations and Flake8 will always run afoul of each other, when it comes to line length.

This is exactly why Rob (and I) had initially suggest we _not_ lint the generated code. If you now agree that linting the migrations is more trouble that it's worth, I'd happily agree. Feel free to submit a PR or comment your approval here and I'll do it.

lparsons commented on DataRepo/models.py:
```diff
@@ -150,13 +150,18 @@ class Protocol(models.Model):
+    # attempt to prevent the loading of duplicate runs, but this does assume
+    # that a distinct sample extract is only run once a day (per
+    # researcher/protocol)
+    unique_together = ("researcher", "date", "protocol", "sample")
```
I think you want to create a [`UniqueConstraint`](https://docs.djangoproject.com/en/3.2/ref/models/constraints/#uniqueconstraint) here.

jcmatese replied:
I can try that. Probably still would not have affect the outcome, because of the current `auto_now_add` setting.

jcmatese replied:
did that

jcmatese commented:
> @jcmatese Looks like there is a missing migration (forgot to check it in perhaps?)
> 
> > I give up on the linting issues, because Django migrations and Flake8 will always run afoul of each other, when it comes to line length.
> 
> This is exactly why Rob (and I) had initially suggest we _not_ lint the generated code. If you now agree that linting the migrations is more trouble that it's worth, I'd happily agree. Feel free to submit a PR or comment your approval here and I'll do it.

Yes, that is fine by me.  I am tired of fighting with it.  Skip migrations green light from me.

jcmatese commented:
The problem is that [pylint thinks this it too long](https://github.com/Princeton-LSI-ResearchComputing/tracebase/runs/2460540100#step:4:791)
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/4b52ec4bfde97600acadb92514989ee4270e36fa/DataRepo/migrations/0007_peakdata_validator.py#L19
But if I edit it to doge that bullet
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/638cc466304d2f75f9404b4a7e6e5fe693e0f655/DataRepo/migrations/0007_peakdata_validator.py#L19-L20
The make migrations complains "that is not what I would have written", presumably.
https://github.com/Princeton-LSI-ResearchComputing/tracebase/runs/2460869637

lparsons commented:
> The problem is that [pylint thinks this it too long](https://github.com/Princeton-LSI-ResearchComputing/tracebase/runs/2460540100#step:4:791)
> https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/4b52ec4bfde97600acadb92514989ee4270e36fa/DataRepo/migrations/0007_peakdata_validator.py#L19
> 
> 
> But if I edit it to doge that bullet
> https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/638cc466304d2f75f9404b4a7e6e5fe693e0f655/DataRepo/migrations/0007_peakdata_validator.py#L19-L20
> 
> 
> The make migrations complains "that is not what I would have written", presumably.
> https://github.com/Princeton-LSI-ResearchComputing/tracebase/runs/2460869637

You removed a space when you made that change.

lparsons commented on requirements/dev.txt:
```diff
@@ -6,3 +6,4 @@ black==20.8b1
 isort
 pylint
 mypy
+openpyxl
```
This belongs in `common.txt` since it's required to actually run the code.

lparsons replied:
Also, I seem to have a missing dependency of `xlrd`. I'm looking into these issues now so I can test.

hepcat72 reviewed (changes requested):
I want to resume my review, but since there were pushes, I don't know what would happen if I refresh this page, so I'm submitted a partial review to not lose my work.

hepcat72 commented:
So 4 of my issues became "outdated" due to the pushes.  The code the issues are about still exists and thus the issues are still relevant.  They are just disconnected.  I'm not sure what the best way is to deal with those disconnected issues.  Probably we should establish a policy to *not push* until all reviewers have responded, but on the grand scale, it seems an unnecessary delay.  Ideally, there would be an easy way to "fix" these "outdated" issues by re-attaching them to the new line numbers.  Unfortunately, there's no mechanism to do that.  (Is that correct?)

The only viable workaround is somewhat labor-intensive: Copy the issue and comments, locate the new location of the code, create a new issue, and paste in the issue content.

Do you guys have an opinion?

jcmatese commented on DataRepo/models.py:
```diff
@@ -209,7 +225,7 @@ class PeakData(models.Model, TracerLabeledClass):
         blank=True,
         validators=[
-            MinValueValidator(1),
+            MinValueValidator(0),
```
yes

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+    def validate_dataframes(self):
+        print("Validating data...")
+
+        # WARNING: I have been told these might vary between Maven versions
+        # columns before sample names expected in original data
+        ORIGINAL_COLUMN_NUMBER = 14
+        # columns before sample names expected in corrected data
+        CORRECTED_COLUMN_NUMBER = 2
```
I will defer to @lparsons on the solution to this issue, regarding the "best practice" for dealing with Maven and AccuCor column-detection (because I imagine he had to deal with it for Maven consumption within AccuCor, and knows about any possible variability in AccuCor headers).  I don't think we can rely on sample name formatting patterns, and there are many other columns that are entirely numeric (medMx, medRt, parent, ppmDiff, etc.) that are not samples.  Lance, is 'parent' always present, and always immediately precedes the samples?

lparsons replied:
The solution I came up with for AccuCor has seemed to work reasonably well. The basic idea is to have a list of column names that are _not_ sample columns (_i.e._ columns that are statically output by Maven / El-Maven / AccuCor). Any columns that are not in that list are considered "sample columns". I believe I lowercased all of the column names prior to matching. The code is here: https://github.com/XiaoyangSu/AccuCor/blob/43d78f9f7ab8615a208f2321491bfb8059f160cc/R/io_functions.R#L135.

I'm OK if we go with this for now and improve it later, if/when we need to. We may be able to simply create a function to return these two values and encapsulate the logic that way.

hepcat72 replied:
I would be fine with a resolution to review issue 1 being the creation a github issue.

hepcat72 replied:
> I added code to compute the minimum sample index, given known/coded columns

Could you please push that change so I can do a complete second round review?  I do not see it.

jcmatese replied:
I think you are not looking at the latest committed fileset?

https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/7665876daa10fbb56cc795e9b46bcf03f1a79818/DataRepo/utils.py#L235-L245

hepcat72 replied:
Yup, you're right. I apparently clicked on a file link that was in an issue attached to "outdated" code and didn't notice that it was an older set of commits I was viewing.  I was attempting to try and determine what was done when there was no comment on the issue, and left that window open.  When I reloaded, it just reloaded that old commit set.  Let me march back through and resolve the issues I commented on after looking at the actual state.

lparsons commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+    """
+
+    def __init__(self, **kwargs):
+        self.accucor_original_df = kwargs.get("accucor_original_df")
+        self.accucor_corrected_df = kwargs.get("accucor_corrected_df")
+        self.date_input = kwargs.get("date")
+        self.protocol_input = kwargs.get("protocol_input").strip()
+        self.researcher = kwargs.get("researcher")
+        self.debug = False
```
I find it a bit odd to strip spaces from command line arguments. It would take some effort to get them there, presumably, they would be intentional? Or is there some other source of this information? Regardless, being consistent would make sense.

jcmatese replied:
I just did it for protocol because that was the argument most likely to be quoted (an extended protocol name/specification).  I can add it to all.

hepcat72 replied:
I didn't realize these were command line arguments.  I just assumed that with the existing strip that they were parsed data.  So I was just going with the safe/cautious assumption.

I will edit the comment to make it NB.  Feel free to resolve without action.

lparsons commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+class AccuCorDataLoader:
+
+    # sample names we are skipping
+    SKIPPED_SAMPLE_NAMES = ["blank"]
```
My suggestion would be to leave out the logic to skip things for now since we did not define any way to identify blank samples by name (we used the value for `Tissue` instead). I'm happy to add that ability in later if/when we define some rules for that (which we could then put into the Sample class and share between scripts).

jcmatese replied:
`SKIPPED_SAMPLE_NAMES` referring code deleted.  Example data file will likely no longer load.

lparsons replied:
@hepcat72 If you approve #58 we can merge that into `main` and this branch and the tests should start passing.

hepcat72 replied:
OK. I will divert from my processing of comments to my review items on this PR to look at #58, then come back.

hepcat72 replied:
Although there's a comment that there was a change, no change is evident in the files changed tab.

lparsons commented:
> So 4 of my issues became "outdated" due to the pushes. 

I seem to see the comments and the code in the "Conversation" tab. I don't think it will be an issue to address them as is, but I'll let @jcmatese have a go at things first. Once this round of comments is addressed we can reassess, but I think this is pretty close.

jcmatese commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+        )
+        parser.add_argument("--date", type=str, help="MSRun date performed")
+        parser.add_argument(
+            "--researcher", type=str, help="Name or identifier of the researcher"
```
Lance already added `required=True`, and I will update the help text

hepcat72 replied:
Same as above.  Required is not set and although there's a comment indicating the help text would change, the change appears to not have been pushed.

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+        except ValueError:
+            return False
+
+    def retrieve_samples(self):
+
+        print("Checking samples...")
+        # cross validate in database
+        self.sample_dict = {}
+        for original_sample_name in self.original_samples:
+            if original_sample_name in self.SKIPPED_SAMPLE_NAMES:
+                continue
+            try:
+                # cached it for later
+                self.sample_dict[original_sample_name] = Sample.objects.get(
+                    name=original_sample_name
+                )
+            except Sample.DoesNotExist as e:
+                print(f"Could not find sample {original_sample_name}")
+                raise (e)
```
B.  There's no check in here about the per-file uniqueness of the sample names, which was mentioned as a requirement in the issue (although perhaps I inferred that - either way, I think it probably should be checked).

jcmatese replied:
does anyone know how `pandas` even reacts to files containing duplicate columns?

jcmatese replied:
or maven/accucor, for that matter?

lparsons replied:
Off the top of my head, I don't know, though I'm sure there are various options on how to react to duplicate column names. Since there is a unique constraint on `MSRun` of  `sample`, `researcher`, 'protocol`, and `date` that should take of any duplicate columns (the insert would fail).

hepcat72 replied:
If that's the case (a DB constraint would catch this), then I think writing a test and including a catch in the code for that exception would be the solution here.

hepcat72 replied:
- [x] Created an issue

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+        if kwargs.get("debug"):
+            self.debug = kwargs.get("debug")
+
+    def validate_data(self):
+        """
+        basic sanity/integrity checks for the data inputs
+        """
+
+        self.validate_dataframes()
+
+        # determine the labeled element from the corrected data
+        self.set_labeled_element()
+
+        self.date = datetime.strptime(self.date_input, "%Y-%m-%d")
+
+        self.retrieve_samples()
+
+        self.retrieve_protocol()
+
+        # cross validate peak_groups/compounds in database
+        self.validate_peak_groups()
```
B. This is an overall design/requirements issue, but it seems like a big one to me.  There is no mitigation of researcher name variants and mispellings, e.g. "Mike" vs. "Michael", with/without first/last/middle name, typos, etc.  Since searching by researcher seems to be an important feature, this is what I think we should do: search for MSRun records (limit 1) containing the researcher name supplied.  If there are 0 records, issue a warning such as "Are you sure you want to insert a new researcher?" - and optionally show a list of distinct researcher names.  You could either name it an interactive thing, or create a hidden flag like "--new-researcher", which would be printed in the warning to allow them to rerun and skip the check.

jcmatese replied:
I am going to pass on this, because it should probably be modeled in the database as an explicit foreign key to a researcher table, eventually.

hepcat72 replied:
Got it.  Although I know I wouldn't trust myself, in terms of typos, to not create name variants.  This is a command line option, not a file content issue, so it has to be manually typed out on every run.  I feel that since search is one, if not the most important, of features for this project, it must be dealt with.  The success of this project could almost be solely measured by the accuracy of the search results.  That said, I think that, at the least, creating an issue to deal with it in the future, whether that's to add a check in this code, or a model change to create a researcher table, should be defined before we merge this PR.

hepcat72 replied:
- [x] created an issue

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+
+        print(f"{action} protocol {self.protocol.id} '{self.protocol.name}'")
+
+    def load_data(self):
```
B.  It's possible I missed this or was out of the loop on model changes, but I couldn't find a check to enforce the (temporary?) requirement that a record with a unique sample, researcher, date, & protocol already exists.  My understanding was that we were going to enforce that as a requirement.  Could you point me to that check if I missed it, or will that happen automatically given field constraints in the DB?  If the latter, perhaps we should gracefully catch that here in the code?

jcmatese replied:
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/5972d57cd7822a4a51409a7666aed9f2f3f35709/DataRepo/models.py#L164-L170

hepcat72 replied:
So it's the latter.  Do you have a catch for the exception that would occur in that instance?  Can you point me to that in the PR (which, I apologize, is what I meant - I should have been more precise).

BTW - heading out to pick up my car. Be back in a bit.

hepcat72 replied:
Yeah, I think you're right.  Thanks.

hepcat72 reviewed (changes requested):
Finishing up my initial review.  Note, I haven't read yet any comments on the first half of my review, so please point out any possible redundancies.  I have a number of requested changes.  One last change request:

We need tests for each of the requirements in the issue:
1. xlsx file (accucor format, required) `test that supplying the wrong file type or no file fails`
2. date required `test that fails when no date supplied`
3. protocol either a integer (protocol.id) or name `test supplying protocol name succeeds` `test that supplying protocol ID succeeds` `test that not supplying a protocol fails`
4. researcher name `test for failure when no name supplied`
5. sample names are unique with the file `supply a file with a duplicate sample name & test for failure`
6. sample names must already be in the database `test for failure with a sample name not in the DB`
7. compound labels must already be in the database `test for failure with a compound not in the DB`

jcmatese commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--accucor_filename", type=str, help="Accucor data xlxs output filepath"
```
Just documenting that superficially, changes on the command-line to `--accucor-file` get retrieved from the parsed options as `accucor_file`
which I find inconsistent and confusing, itself?  But adding it...

calling this
```
$ python manage.py load_accucor_msruns --accucor-file "DataRepo/example_data/obob_maven_6eaas_inf.xlsx" --protocol 1 --date "2021-04-23" --researcher "mneistat" --debug
```

and `print(options)` during run.

```
{'verbosity': 1, 'settings': None, 'pythonpath': None, 'traceback': False, 'no_color': False, 'force_color': False, 'skip_checks': False, 'accucor_file': 'DataRepo/example_data/obob_maven_6eaas_inf.xlsx', 'protocol': '1', 'date': '2021-04-23', 'researcher': 'mneistat', 'debug': True}
```

hepcat72 replied:
Yes, they do.  Sorry.  I should have offered that.  Lance explained it to me when I'd brought up the underscore/dash issue in another PR.

hepcat72 replied:
As far as I can see, blocking issue 1 was not implemented, although a comment was added saying it would be done.  Is there a missing push?

hepcat72 replied:
Also, required is not set here.  Have Lance's changes not been merged into this branch?  I think his changes should be incorporated.

jcmatese replied:
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/7665876daa10fbb56cc795e9b46bcf03f1a79818/DataRepo/management/commands/load_accucor_msruns.py#L14-L46

lparsons commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+        for original_sample_name in self.original_samples:
+            if original_sample_name in self.SKIPPED_SAMPLE_NAMES:
+                continue
+            try:
+                # cached it for later
+                self.sample_dict[original_sample_name] = Sample.objects.get(
+                    name=original_sample_name
+                )
+            except Sample.DoesNotExist as e:
+                print(f"Could not find sample {original_sample_name}")
+                raise (e)
```
Seems like a nice to have. I may regret it one day, but I'd say this can wait until we actually need it.

hepcat72 replied:
That's what I was thinking, thus the `NB`.

hepcat72 replied:
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+        """
+        print(f"Finding or inserting protocol for '{self.protocol_input}'...")
+
+        action = "Found "
+
+        if self.is_integer(self.protocol_input):
+            try:
+                self.protocol = Protocol.objects.get(id=self.protocol_input)
+            except Protocol.DoesNotExist as e:
+                print("Protocol does not exist.")
+                raise (e)
+        else:
+            try:
+                self.protocol, created = Protocol.objects.get_or_create(
+                    name=self.protocol_input
+                )
+
+                if created:
+                    action = "Created "
+            except Exception as e:
+                print(f"Failed to get or create protocol {self.protocol_input}")
+                raise e
+
+        print(f"{action} protocol {self.protocol.id} '{self.protocol.name}'")
```
modified

hepcat72 replied:
Although this is stated as changed, no change is evident.

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+                self.validate_data()
+                self.load_data()
+        except (IntegrityError, Exception) as e:
+            print(e)
+            raise (e)
```
I modified some of the code I added, but will not be making sitewide edits, and I almost certainly copied `raise (e)` from somewhere else in the code base

hepcat72 replied:
Fair enough.

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+                    #  PeakGroup
+                    for compound in peak_group_attrs["compounds"]:
+                        peak_group.compounds.add(compound)
+
+                # we should have a cached PeakGroup now
+                peak_group = inserted_peak_group_dict[peak_group_name]
```
I am not certain about what a duplicate peak group name would be, exactly, in the context of these files.  Certainly, compound is repeated multiple times in the file (once for every isotope), so the only concept of a "duplicate" might be a duplicated row, perhaps, or maybe the same compound name used for different formulas/medMz/medRt or some other data inconsistency?

hepcat72 replied:
Well, maybe answering this question would provide the answer: Why did you make this check on line 416?:

```
if peak_group_name not in inserted_peak_group_dict:
```

hepcat72 replied:
Because there's no else, and no warning about skipped data, I was guessing that you had a reason for the check and no `else` code.  If we're skipping the existing peak groups, I'd add the answer in a comment.

jcmatese replied:
I think the logic boils down to 


```
# have we already inserted and cached this peak group?
if peak_group_name not in inserted_peak_group_dict:
      # it is not is the cache, so insert it and cache it

# it should be inserted and cached, now
peak_group = inserted_peak_group_dict[peak_group_name]
```

hepcat72 replied:
Oh wait.  I think I get it.  Is this right?  A peak group recurs in the file because multiple rows have things that belong to it.  You are more versed in the data relationships at that level than I am.

jcmatese replied:
yes, it is the labeled_count that changes, per line.  So this "corrected" excerpt would be 1 peak_group, for 1 sample

![Screen Shot 2021-04-30 at 7 05 26 PM](https://user-images.githubusercontent.com/6091114/116762007-22b8ad80-a9e7-11eb-86f9-2066c0b13b21.png)

You pretty much have to take in the entire nested loop(s), to get the full context

https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/b803989f8a7c963a125ff7eda6d375510c53e18b/DataRepo/utils.py#L396-L442

hepcat72 replied:
OK, now that I understand better, I have a subsequent question & 2 issues...

Q: The peak group name is essentially the distinct values in the "compound" column, do we need to worry about different compounds having separate rows but in the same peak group?  Perhaps you already took that into account.  And what was the decision on case variations?  Is that already handled?

B: The primary purpose of tracebase is to be able to go back and extract unanalyzed peak groups.  In the process of retrieving, analyzing, and loading these peak groups, we can clearly expect to encounter the same MSRun (with the same date, researcher, protocol, and sample) as was previously loaded.  However, the code from 402-411 will most assuredly cause an exception given the unique constraint and the new data will fail to load.  There appears to be no check on the record as pre-existing, nor a retrieval of that MSRun record to be able to add new peak groups that belong to it.  This seems to be a critical error to me.

NB: I also think that we're going to have an issue catching cases where a user has either mistakenly or intentionally included the same peak group from separate El-Maven/AccuCor of the same sample (a use-case we explicitly do not support).  And I think that if we fix the above issue about about existing MSRun entries, the load will fail due to existing peak groups because there is no check on existing peak groups.  You could make the case that this is what we want, but I just would like that to be an intentional decision.  Do we want to not load any data from the file if there is an existing peak group in the database or just warn or ignore existing peak groups?

hepcat72 replied:
- [x] created an issue

lparsons commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+
+    def load_accucor_data(self):
+
+        try:
+            with transaction.atomic():
+                self.validate_data()
+                self.load_data()
+        except (IntegrityError, Exception) as e:
+            print(e)
+            raise (e)
```
The reason this isn't caught is that this isn't a typical use of parens. Since `raise` isn't a funtion, but instead a keyword, the parenthesis in this case are simply being ignored. They are often used to [break up long lines](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). In this case, the typical syntax would be simply:
```python
raise e
```

While we are on these lines, `except (IntegrityError, Exception) as e:` is rather odd as well. This catches _all_ `Exception`s so adding the special case for the `IntegrityError` is not necessary. Also, I don't think there is any need to both print and raise the exception since. In fact, there is no need for the try at all since the Exception is being raised.

The same effect could be had with:
```python
        with transaction.atomic():
            self.validate_data()
            self.load_data()
```

hepcat72 replied:
WRT the existence of the `Exception` class in the catch, I ran afoul of some Python experts (I think on stack) who admonished me for using `Exception` in a catch at all.  There were some cases where it was acceptable.  But in this case, if there is something you want to do WRT the specific integrity exception, that should be in a separate catch above.  (Although I think you may be able to switch on the type inside a single catch.  In either case,)  here, you're doing the same thing regardless of the type, so the integrity one is unnecessary - and like Lance said, you're only raising, so you don't need the try/catch at all.

lparsons replied:
> essentially, that `try` wraps a host of other `try`'s, each potentially with their own exception type, but I want them **_all_** to rollback, regardless.

That makes sense. But there is no need to "catch" the exception just to raise it again. It would only be useful if you were going to do something with it (like roll back some previous transaction, skip to the next row, etc.).

jcmatese replied:
I guess I would have normally have put an explicit `rollback` in that `except` block , but I don't really know how Django handles it.  I can strip those away, and hope for the best.

lparsons replied:
> I guess I would have normally have put an explicit `rollback` in that `except` block , but I don't really know how Django handles it. I can strip those away, and hope for the best.

I'm not certain how Django handle's things either, I'd have to read the Docs on the transaction stuff. Having a test for this would be make sure we didn't mess this up though.

jcmatese replied:
removed `try: except:` from `load_accucor_data`

lparsons commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+
+    # determine the labeled element from the corrected data
+    def set_labeled_element(self):
+        label_pattern = "^([CNHOS]{1})_Label$"
```
Blocking: Actually, this seems unnecessary and duplicates the validation checks already built into the model. That means we have two checks which could get out of sync and would have to be maintained separately. Instead, [calling [`full_clean()`](https://docs.djangoproject.com/en/3.2/ref/models/instances/#validating-objects) on an object will ensure it's valid. If not, it will throw and exception, which will bubble up until it's "caught" (or not). 

In this case, since you've used [atomic transaction](https://docs.djangoproject.com/en/3.2/topics/db/transactions/#django.db.transaction.atomic), there is no need to do any explicit checks that exist in the database models since any exceptions will cause all of the transactions to not get committed. At least, that is what I believe the intent of the code is.

hepcat72 replied:
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+
+                if peak_group_name not in inserted_peak_group_dict:
+
+                    # here we insert and cache a PeakGroup, by name (only once
+                    # per file). NOTE: if the first row encountered has any
+                    # issues (for example, a null formula, as sometimes observed
+                    # in original Maven results), then this block will likely
+                    # fail
```
I was just making an observation on row 38 of [this example file](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/9c2df67183cd96665908889da0a5fbf331fb3d6c/DataRepo/example_data/obob_maven_6eaas_inf_original.tsv)
![Screen Shot 2021-05-03 at 10 27 51 AM](https://user-images.githubusercontent.com/6091114/116889308-67686280-abfa-11eb-8c50-433221cae0ce.png)

That row, presumably a peak group with no measured data, has no formula, medMz or medRt.  If that was encountered as the first row `C12 PARENT` instead of the `C13-label-4` isotopomer, I imagine there will be issues, because PeakGroup.formula is currently required (AFAIK).  Res-sorting, as you suggest, may mitigate this. Also, have not yet checked to see what  happens to those empty cells when inserted PeakData.

hepcat72 replied:
OK. Let's suss this out.  I'm not actually understanding how this will fail and I would like to understand the issue(s) alluded to in the comments so that they can be clarified.  The formula field in PeakGroup doesn't have any requirements.  It can be null, empty string, and non-unique.  The associated compound is retrieved by name and will have its own formula.

NB: Vague comment about unspecified issue WRT empty formula.

If the issue alluded to in the comments is not a defined issue in existing code or in the code of the PR, but a future issue is foreseen/predicted, perhaps this PR should include a constraint on the PeakGroup.formula field?  If we're aware of it now, let's nail it down so that we don't spin our wheels later trying to re-figure it out / discover it again out of context.

hepcat72 replied:
- [ ] Write a test that shows it fails

hepcat72 replied:
Please check the box on the comment above when the test is implemented.

lparsons replied:
This is essentially validating data that gets inserted into the database. I believe it's better design to have the validation done at the lower level in the model so any other method of updating or inserting data would also trigger it. We know we want to ensure that `PeakGroup.formula` and `Compound.formula` have validation that ensure they are not empty and can be parsed by [`chempy.Substance.from_formula`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/60/files#diff-a2bf8b389498d194316674213a4f1dcf1244f93a9e8c2dc0831a1008d9a64b8aR49). I'll create an issue for implementing these validations and we can discuss any other validations we want there.

jcmatese replied:
the formula validation might be best served during model save() somehow (perhaps using chempy or cross-check with compounds). We might also tweak the incoming data with pandas read_excel dtype or converters arguments, which might set default behavior for null cells/strings. The jury is still out on the best solution, and Lance proposes #61

hepcat72 replied:
I think there are a few issues here, some related to this PR and others related to the model.  I think it's valid to expect an exception from the database due to the alluded to problem in this comment:

https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/68469624485631fe8624d83a7090c2cccfe7b2c7/DataRepo/utils.py#L485-L488

I also think it's valid to not implement the test (since we know that that exception wouldn't currently happen).

There still remains the issue of the vague comment implying likely failure, however I will add an item to the description of #61 that is to change the comment when we do know what will happen.

Resolving.

hepcat72 replied:
So does this mean that we will keep the empty formula row?  I noted that issue #61 does not specify the implementation of tests, so I will add that too.

jcmatese replied:
The data are produced by the Maven+Accucor pipeline, unless the user goes in and edits the file (which they really should not).  I don't know why the formula is not written, but I suspect it is because the peak_group/row measures **0 for every sample**. [See line 38, here](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/DataRepo/example_data/obob_maven_6eaas_inf_original.tsv#L38) , an scroll across.  When I noticed this anomaly, I made the comment out of concern, because I did not know how the would data possibly would transit over the various layers (pandas through Django ORM).  When I edit the file, and delete the formula from a [C12 PARENT](url) line and inspect it by printing the formula, it reports as `nan`, and so I expect that might be the eventual formula value, if this were to ever happen.  File testing seems to bear this guess out

```
{'_state': <django.db.models.base.ModelState object at 0x7ffcf8638880>,
 'formula': 'nan',
 'id': 7470,
 'ms_run_id': 1068,
 'name': 'glucose'}
```

I have tried to be transparent about everything, and document the best I can.

hepcat72 replied:
You did. WRT the vagueness of this comment, the vagueness is baked-in (i.e. not due to lack of effort on your part), because there's no way to know what would happen, and that's because there are no constraints on the field in the database.  In fact, I think your comment improved the overall design/requirements, because it highlighted a flaw in the model.

jcmatese replied:
Just documenting that _if_ you did want to put the constraint logic in the loader class, instead of in the PeakGroup model class, you _could_ probably warn or throw an exception with

```
            if pandas.isnull(row["formula"]):
                print(f"Warning: {peak_group_name} formula is null!")
```

hepcat72 replied:
Honestly, that would probably be useful since the database constraint exception may(?) not provide any direction as to where in the file the data was that caused the exception.

fkang-pu commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+        basic sanity/integrity checks for the data inputs
+        """
+
+        self.validate_dataframes()
+
+        # determine the labeled element from the corrected data
+        self.set_labeled_element()
+
+        self.date = datetime.strptime(self.date_input, "%Y-%m-%d")
+
+        self.retrieve_samples()
+
+        self.retrieve_protocol()
+
+        # cross validate peak_groups/compounds in database
+        self.validate_peak_groups()
```
I agree with Rob. It would be better to go through the whole validation process and list all failed items in the log, then raise an exception for loading if the number of failed item(s)>1.

hepcat72 replied:
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

jcmatese replied:
passing on this for now

fkang-pu reviewed (commented):
The transaction.atomic seems working. I checked out this pull request yesterday morning, and was able to load data from obob_maven_6eaas_inf.xlsx (after fixing row for "citrate" in compound table), but got error for loading "obob_maven_c160_inf.xlsx" due to missing sample "bkbk1123". 

One concern I have is we only store accucor file name(s) in the log for msrun/peakdata loading. Querying database after loading, we wouldn't know its source file(s) for a specific peakgroup.  I think it's important to be able to trace back to the source file, which would be helpful for data verification and search/grouping. My proposal is adding dataset and archived_file tables, or at least dataset for now (peakgroup has a foreign key to dataset, dataset has M:M relationship to archive_file). Not sure if it makes sense to you. I raise the question without approving this pull request, as it could affect Peakgroup model structure and loading script. 

p.s. could  add code to check formula in addition to verify/link peakgroup to compound(s). make sure the formula for a peakgroup matches that in compound table.

hepcat72 commented:
Let me know when this is ready for a second-round review.

hepcat72 commented:
> Finishing up my initial review. Note, I haven't read yet any comments on the first half of my review, so please point out any possible redundancies. I have a number of requested changes. One last change request:
> 
> B. We need tests for each of the requirements in the issue:
> 
> 1. [ ]  xlsx file (accucor format, required) `test that supplying the wrong file type or no file fails`
> 2. [ ]  date required `test that fails when no date supplied`
> 3. [ ]  protocol either a integer (protocol.id) or name `test supplying protocol name succeeds` `test that supplying protocol ID succeeds` `test that not supplying a protocol fails`
> 4. [ ]  researcher name `test for failure when no name supplied`
> 5. [ ]  sample names are unique with the file `supply a file with a duplicate sample name & test for failure`
> 6. [ ]  sample names must already be in the database `test for failure with a sample name not in the DB`
> 7. [ ]  compound labels must already be in the database `test for failure with a compound not in the DB`

- [x] Lance created an issue for these

jcmatese commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+    def set_labeled_element(self):
+        label_pattern = "^([CNHOS]{1})_Label$"
+        labeled_df = self.accucor_corrected_df.filter(regex=(label_pattern))
+        # lets hope sample names don't collide with AccuCor column labeling
+        assert (
+            len(labeled_df.columns) == 1
+        ), "Loader cannot deal with multiple labels, currently..."
```
will print out header and remove comment

hepcat72 replied:
This change, I could actually see! Thanks.

hepcat72 commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+class Command(BaseCommand):
+    # Show this when the user types help
+    help = (
+        "Loads data from an accucor file table into"
```
This was marked as resolved without actually implementing the change.  Either way, there should be a comment, but especially if the change is rejected. Since there is no indication of what the resolution is, I am going to mark it as unresolved.  Please address issues in the following manner:

1. Comment to indicate your resolution (whether it's rejected or accepted) and (if NB: optionally) stating either the reason for the rejection of what was done.  If B, do not resolve either way.
2. If accepted and the suggestion was implemented as suggested, resolve.  If rejected or an alternate resolution was made, leave as unresolved.

Following this process makes the review process less labor intensive and run more smoothly.

hepcat72 replied:
If the intention was to reject this suggested change or if it was implemented but not yet pushed, feel free to re-resolve.  But again, please always comment on each issue so reviewer don't have to guess what the resolution was.

jcmatese replied:
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/7665876daa10fbb56cc795e9b46bcf03f1a79818/DataRepo/management/commands/load_accucor_msruns.py#L10-L11

hepcat72 commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+        parser.add_argument(
+            "--protocol", type=str, help="protocol database identifier or name"
+        )
+        parser.add_argument("--date", type=str, help="MSRun date performed")
```
I do not see the commented changes.

hepcat72 commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+            "--debug",
+            action="store_true",
+            default=False,
+            help='Raise a "debug-only" error, to abort transaction',
```
No resolution offered for this item and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

jcmatese replied:
https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/7665876daa10fbb56cc795e9b46bcf03f1a79818/DataRepo/management/commands/load_accucor_msruns.py#L44-L45

hepcat72 commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+        )
+
+    def handle(self, *args, **options):
+        print("Reading accucor file(s)")
```
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

hepcat72 commented on DataRepo/management/commands/load_accucor_msruns.py:
```diff
@@ -0,0 +1,50 @@
+
+
+class Command(BaseCommand):
+    # Show this when the user types help
```
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

hepcat72 commented on DataRepo/models.py:
```diff
@@ -150,13 +150,24 @@ class Protocol(models.Model):
 
+    # attempt to prevent the loading of duplicate runs, but this does assume
+    # that a distinct sample extract is only run once a day (per
+    # researcher/protocol)
```
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+    def validate_peak_groups(self):
+
+        # step through the original file, and note all the unique peak group
+        # names/formulas and map to database compounds
```
No resolution offered for this item, no change evident, and given the previously noted commented changes that are not evident in the code, I have no idea whether this was intended to be changed or whether it was rejected.

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -184,3 +192,280 @@ def load_sample_table(self, data):
+
+                for compound_input in compounds_input:
+                    try:
+                        # cache it for later; note, if the firs row encountered
```
Although this is stated as changed, no change is evident.

hepcat72 reviewed (changes requested):
There were many issues marked as resolved for which no resolution or explanation was offered.  And where a comment was added indicating a change, there is no change evident in the code.  I do not understand why this is the case, but I can only assume that the changes have not been pushed.

Let me re-affirm the order of operations I thought we were all following:

1. Submit a PR & request an initial review
2. Reviewers add issues
3. Author accepts/rejects each issue
4. Author implements changes to the accepted issues
5. Author pushes that work to the PR branch
6. Author requests subsequent review
7. Reviewer checks all issues, evaluates the changes made or the reason for rejecting each issue and either agrees to the proposed resolution (and resolves) or starts a dialog on the issue.  New issues may be created for new code & the review is submitted.
8. After all reviewers have finished their reviews, if changes are requested or there or further comments are requested, go back to step 3.  Otherwise, merge.

I was also expecting that Lance's changes would be merged with this branch and that all existing changes on this branch had been pushed.

Is there a technical issue here?  Why do I not see the changes that were commented to have been done?

hepcat72 commented on DataRepo/models.py:
```diff
@@ -37,6 +37,22 @@ class TracerLabeledClass:
     MAX_LABELED_ATOMS = 20
 
+    @classmethod
+    def tracer_labeled_element_regex_pattern(cls):
```
NB. This doesn't return a regex pattern.  It returns the string `CHNOS` (or some other order of characters).  I suggest a more accurate name:
```suggestion
    def tracer_labeled_elements_concat_string(cls):
```

jcmatese replied:
maybe I will move this method/function to the Loader class, instead.  This method predated the other method I added (`tracer_labeled_elements_list`) , so unless anyone thinks they "will also need a concatenated string", I might go that route.

jcmatese replied:
I modified/moved it

hepcat72 commented on DataRepo/models.py:
```diff
@@ -37,6 +37,22 @@ class TracerLabeledClass:
+    def tracer_labeled_element_regex_pattern(cls):
+
+        tracer_element = ""
+        for idx in cls.TRACER_LABELED_ELEMENT_CHOICES:
+            tracer_element = tracer_element + idx[0]
+
+        return tracer_element
```
B. This code is fragile to future changes because it assumes that the element symbol is a single character.  I think the code is potentially fine, as long as the assumption is documented both here and in the `TracerLabeledClass` so that a coder potentially adding elements can catch any potential issue.

jcmatese replied:
maybe I will move this method/function to the Loader class, instead.  This method predated the other method I added (`tracer_labeled_elements_list`) , so unless anyone thinks they "will also need a concatenated string", I might go that route.

jcmatese replied:
I modified/moved it

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -233,16 +231,22 @@ def validate_data(self):
-        ORIGINAL_COLUMN_NUMBER = 14
-        # columns before sample names expected in corrected data
-        CORRECTED_COLUMN_NUMBER = 2
+        # column index of the first predicted sample for the original data
+        original_minimum_sample_index = self.get_first_sample_column_index(
+            self.accucor_original_df
+        )
+        # column index of the first predicted sample for the corrected data
+        corrected_minimum_sample_index = self.get_first_sample_column_index(
+            self.accucor_corrected_df
+        )
```
C. Great solution!

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -296,25 +310,71 @@ def is_integer(self, data):
+            "Compound",
+        ]
+
+        # append the *_Label columns of the corrected dataframe
+        tracer_element_list = TracerLabeledClass.tracer_labeled_elements_list()
+        for element in tracer_element_list:
+            NONSAMPLE_COLUMN_NAMES.append(f"{element}_Label")
```
C. Nice touch.

hepcat72 reviewed (changes requested):
There is 1 new non-blocking issue, 1 new blocking (but trivial to resolve) issue, 1 old non-blocking issue that I do not know what the intended resolution to was, and 1 old blocking issue.  I can't link the new issues in this comment (so I will qualify them), but the old issues are:

- [ ] https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/55#discussion_r624097097
- [ ] https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/55#discussion_r626012690
- [ ] NB. tracer_labeled_element_regex_pattern method name
- [ ] B. Document the assumption about element symbols being a single character (which could possibly not be true in the future).

However the "resolved" statuses of all my issues are current.  If it is or is not "resolved", that status is accurate according to my latest round review.

jcmatese commented:
@hepcat72 I changed the regex, but will not address the other two in this go around.

@lparsons believed the formula validation might be best served during model save() somehow (perhaps using chempy or cross-check with compounds).  We might also tweak the incoming data with pandas `read_excel` `dtype` or `converters` arguments, which might set default behavior for null cells/strings.  The jury is still out on the best solution, and Lance proposes #61
 
related: I did remove the `peak_group.name.lower()` accommodation in commit [4314796](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/55/commits/4314796cbfb3b87f5e488ab7b734641fcf20e895) and updated all the example files to the next of my ability.

regarding reporting all issue in 1-pass, instead of two, I suggest we get this in usage and see how often it is a problem.  I also suspect this will not be the first time this code is modified, and perhaps an entirely different interface might be put on it (multiple tsv files, web-wrapped, etc.)

hepcat72 commented:
Well, re-request my review when you're ready for me to take another look to re-evaluate.  Those 2 outstanding issues (if I know which ones you're referring to) should be ok to pass on given Lance created an issue for the blocking one.

hepcat72 commented:
> regarding reporting all issue in 1-pass, instead of two, I suggest we get this in usage and see how often it is a problem. I also suspect this will not be the first time this code is modified, and perhaps an entirely different interface might be put on it (multiple tsv files, web-wrapped, etc.)

Could you explain what you're referring to here @jcmatese?

jcmatese commented:
> > regarding reporting all issue in 1-pass, instead of two, I suggest we get this in usage and see how often it is a problem. I also suspect this will not be the first time this code is modified, and perhaps an entirely different interface might be put on it (multiple tsv files, web-wrapped, etc.)
> 
> Could you explain what you're referring to here @jcmatese?

https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/55#discussion_r624097097

hepcat72 reviewed (approved):
All issues resolved. 🎉
